File: python/fcn_kaplanMeier.py
# fcn_kaplanMeier.py
import numpy as np
from typing import Tuple

def fcn_kaplanMeier(
    event_times: np.ndarray,
    event_indicators: np.ndarray,
) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """
    Kaplan–Meier survival estimates (with Greenwood SE), matching the MATLAB logic.

    Parameters
    ----------
    event_times : array-like
        Times of event or censoring.
    event_indicators : array-like
        1 if event (death), 0 if censored.

    Returns
    -------
    t_km : (m,) ndarray
        Unique event times (where events occurred).
    s_km : (m,) ndarray
        Survival probability at each event time.
    se_km : (m,) ndarray
        Standard errors via Greenwood's formula.
    n_risk : (m,) ndarray
        Number at risk just prior to each event time.
    """
    # -- sanitize & sort --
    times = np.asarray(event_times, dtype=float).ravel()
    events = np.asarray(event_indicators, dtype=float).ravel()  # allow NaN filter; cast to int later

    valid = ~np.isnan(times) & ~np.isnan(events)
    times = times[valid]
    events = events[valid].astype(int)

    order = np.argsort(times, kind="mergesort")
    times = times[order]
    events = events[order]

    # unique times where an event occurred
    unique_times = np.unique(times[events == 1])
    n_times = unique_times.size

    t_km = unique_times.copy()
    s_km = np.ones(n_times, dtype=float)
    se_km = np.zeros(n_times, dtype=float)
    n_risk = np.zeros(n_times, dtype=float)

    survival_prob = 1.0
    variance_sum = 0.0  # Greenwood accumulator

    for i, t in enumerate(unique_times):
        n_at_risk = np.sum(times >= t)
        n_risk[i] = n_at_risk

        n_events = np.sum((times == t) & (events == 1))

        if n_at_risk > 0:
            survival_prob *= (1.0 - n_events / n_at_risk)
            s_km[i] = survival_prob

            if n_events > 0:
                denom = n_at_risk * (n_at_risk - n_events)
                if denom > 0:
                    variance_sum += n_events / denom
                else:
                    variance_sum = np.inf  # all-at-risk fail at once → infinite variance

            if survival_prob > 0 and np.isfinite(variance_sum):
                se_km[i] = survival_prob * np.sqrt(variance_sum)
            elif survival_prob <= 0:
                se_km[i] = 0.0
            else:
                se_km[i] = np.inf

    return t_km, s_km, se_km, n_risk

File: matlab/fcn_kaplanMeier.m
function [t_km, s_km, se_km, n_risk] = kaplanMeier(event_times, event_indicators)
% KAPLANMEIER Calculate Kaplan-Meier survival estimates
%
% Inputs:
%   event_times     - Vector of event/censoring times
%   event_indicators - Vector of event indicators (1=event, 0=censored)
%
% Outputs:
%   t_km    - Unique event times
%   s_km    - Survival probability estimates at each event time
%   se_km   - Standard errors (Greenwood's formula)
%   n_risk  - Number at risk at each event time
%
% Reference: Klein & Moeschberger (2003), Survival Analysis

% Remove any NaN or missing values
valid_idx = ~isnan(event_times) & ~isnan(event_indicators);
event_times = event_times(valid_idx);
event_indicators = event_indicators(valid_idx);

% Sort by time
[event_times, sort_idx] = sort(event_times);
event_indicators = event_indicators(sort_idx);

% Get unique event times (where actual events occurred)
unique_times = unique(event_times(event_indicators == 1));

% Initialize outputs
n_times = length(unique_times);
t_km = unique_times;
s_km = ones(n_times, 1);
se_km = zeros(n_times, 1);
n_risk = zeros(n_times, 1);

% Calculate KM estimates
survival_prob = 1;
variance_sum = 0;

for i = 1:n_times
    t = unique_times(i);
    
    % Number at risk at time t
    n_at_risk = sum(event_times >= t);
    n_risk(i) = n_at_risk;
    
    % Number of events at time t
    n_events = sum(event_times == t & event_indicators == 1);
    
    % Update survival probability
    if n_at_risk > 0
        survival_prob = survival_prob * (1 - n_events / n_at_risk);
        s_km(i) = survival_prob;
        
        % Update variance sum for Greenwood's formula
        if n_events > 0
            variance_sum = variance_sum + n_events / (n_at_risk * (n_at_risk - n_events));
        end
        
        % Calculate standard error
        if survival_prob > 0
            se_km(i) = survival_prob * sqrt(variance_sum);
        else
            se_km(i) = 0;
        end
    end
end

% Ensure column vectors
t_km = t_km(:);
s_km = s_km(:);
se_km = se_km(:);
n_risk = n_risk(:);

end